src/blockchains/erc20/lib/add_balances.ts

```diff
@@ -162,25 +163,28 @@ function identifyAddresses(events: ERC20Transfer[]): BlockNumberToAffectedAddres
 
 // Build a balance map for all addresses involved in transactions
 async function buildBalancesMap(web3: Web3, batchedAddresses: [number, Utils.AddressContract[]][],
-  maxConnectionConcurrency: number | undefined, multicallAddress: string = MULTICALL_ADDRESS): Promise<BlockNumberToBalances> {
+  maxConnectionConcurrency: number, multicallAddress: string = MULTICALL_ADDRESS): Promise<BlockNumberToBalances> {
   const results: Utils.BlockNumberAddressContractBalance[] = []
+  if (batchedAddresses.length === 0) {
+    return new Map() as BlockNumberToBalances
+  }
+
+  const chunkCount = Math.min(maxConnectionConcurrency, batchedAddresses.length)
+  const chunkSize = Math.ceil(batchedAddresses.length / chunkCount)
+  const indexChunks = buildInclusiveChunks(0, batchedAddresses.length - 1, chunkSize)
 
-  let concurrentRequests: Promise<Utils.BlockNumberAddressContractBalance[]>[] = []
-  let countConcurrent = 0
-  const concurrencyLimit = maxConnectionConcurrency ?? Number.POSITIVE_INFINITY
-
-  for (const blockNumberAddress of batchedAddresses) {
-    concurrentRequests.push(getBalancesPerBlock(web3, blockNumberAddress[1], blockNumberAddress[0], multicallAddress))
-    countConcurrent += 1
-    if (countConcurrent >= concurrencyLimit) {
-      results.push(... (await Promise.all(concurrentRequests)).flat())
```

Comment:
Remove this intermediate wait. Break the whole interval according to parallelism and send at once.